File: tests/layer0_ops/unit/TestSubsampling.m
classdef TestSubsampling < matlab.unittest.TestCase
    % TestSubsampling: unit tests for comp_downs and comp_ups
    
    
    methods(TestClassSetup)
        function setupTestClass(testCase)
            sharedDir = fullfile(fileparts(mfilename('fullpath')), ...
                                 '..', '..', 'shared');
            addpath(sharedDir);
            setup_filterbank_paths();
        end
    end

    methods(Test)
        
        function testCompsDownsIdentity(testCase)
            % comp_downs(x, 1) = x (a=1 is identity)
            rng(42);
            for trial = 1:10
                x = randn(128, 1);
                result = comp_downs(x, 1);
                testCase.verifyEqual(result, x, 'AbsTol', 1e-14);
            end
        end
        
        function testCompsDownsLength(testCase)
            % comp_downs with a=2: output length = ceil(L/2)
            rng(42);
            for trial = 1:10
                L = randi([100, 500]);
                x = randn(L, 1);
                a = 2;
                result = comp_downs(x, a);
                expected_len = ceil(L / a);
                testCase.verifyEqual(length(result), expected_len);
            end
        end
        
        function testCompsDownsUpsCyclic(testCase)
            % comp_downs followed by comp_ups recovers impulse structure (check specific positions)
            rng(42);
            % Create impulse
            x = zeros(128, 1);
            x(1) = 1;
            
            a = 2;
            down = comp_downs(x, a);
            up = comp_ups(down, a);
            
            % The up-sampled version should have impulse at positions 1, 1+a, 1+2a, ...
            testCase.verifyTrue(up(1) > 0, 'Impulse lost after down/up cycle');
        end
        
        function testCompsUpsIdentity(testCase)
            % comp_ups(x, 1) = x (a=1 is identity)
            rng(42);
            for trial = 1:10
                x = randn(128, 1);
                result = comp_ups(x, 1);
                testCase.verifyEqual(result, x, 'AbsTol', 1e-14);
            end
        end
        
        function testCompsUpsLength(testCase)
            % comp_ups output length = a * length(x)
            rng(42);
            for trial = 1:10
                x = randn(64, 1);
                a = randi([2, 8]);
                result = comp_ups(x, a);
                testCase.verifyEqual(length(result), a * length(x));
            end
        end
        
        function testCompsUpsEnergy(testCase)
            % Energy check: sum(comp_ups(x,a).^2) = sum(x.^2) (upsampling preserves energy)
            rng(42);
            for trial = 1:10
                x = randn(64, 1);
                a = randi([2, 8]);
                up = comp_ups(x, a);
                energy_in = sum(abs(x).^2);
                energy_out = sum(abs(up).^2);
                testCase.verifyEqual(energy_out, energy_in, 'RelTol', 1e-12);
            end
        end
        
        function testCompsDownsSkipParameter(testCase)
            % skip parameter: comp_downs(x, a, skip) starts at index 1+skip
            rng(42);
            x = randn(256, 1);
            a = 2;
            skip = 3;
            
            result_no_skip = comp_downs(x, a);
            result_skip = comp_downs(x, a, skip);
            
            % With skip, we start at x(1+skip), so first output should match skip version
            % (verification is qualitative - just check it runs without error)
            testCase.verifyEqual(length(result_skip), length(result_no_skip), ...
                'RelTol', 0.1);  % Allow some length difference due to skip
        end
        
        function testCompsDownsZeroSignal(testCase)
            % Test with zero signal
            x = zeros(128, 1);
            result = comp_downs(x, 2);
            testCase.verifyEqual(result, zeros(64, 1), 'AbsTol', 1e-14);
        end
        
        function testCompsDownsImpulse(testCase)
            % Test with impulse signal
            x = zeros(128, 1);
            x(1) = 1;
            result = comp_downs(x, 2);
            testCase.verifyEqual(result(1), 1, 'AbsTol', 1e-14);
        end
        
        function testCompsDownsNoise(testCase)
            % Test with real noise
            rng(42);
            x = randn(256, 1);
            a = 2;
            result = comp_downs(x, a);
            testCase.verifyEqual(length(result), ceil(256/a));
        end
        
    end
    
end

File: tests/layer0_ops/unit/test_subsampling.py
"""
test_subsampling.py
===================
Python port of:
    layer0_ops/unit/TestSubsampling.m

Covers: comp_downs, comp_ups
"""

from __future__ import annotations

import pytest

import numpy as np
from conftest import comp_downs_ref, comp_ups_ref

# ---------------------------------------------------------------------------
# comp_downs – reference tests
# ---------------------------------------------------------------------------

class TestCompDownsReference:
    """
    MATLAB counterpart: TestSubsampling (comp_downs section).
    """

    def test_identity_a1(self):
        """comp_downs(x, 1) == x (a=1 is identity)."""
        rng = np.random.default_rng(0)
        for _ in range(10):
            x = rng.standard_normal(128)
            np.testing.assert_array_equal(comp_downs_ref(x, 1), x)

    def test_output_length(self):
        """comp_downs(x, 2) has length ceil(L/2)."""
        rng = np.random.default_rng(1)
        for _ in range(10):
            L      = rng.integers(100, 501)
            x      = rng.standard_normal(int(L))
            result = comp_downs_ref(x, 2)
            assert len(result) == int(np.ceil(L / 2))

    def test_zero_input(self):
        """comp_downs of a zero vector is zero."""
        x = np.zeros(128)
        np.testing.assert_array_equal(comp_downs_ref(x, 2), np.zeros(64))

    def test_impulse(self):
        """comp_downs of an impulse at position 0 preserves that sample."""
        x    = np.zeros(128); x[0] = 1.0
        out  = comp_downs_ref(x, 2)
        assert out[0] == 1.0

    def test_noise_output_length(self):
        """comp_downs(noise, 2): output length = ceil(L/2)."""
        L   = 256
        x   = np.random.default_rng(2).standard_normal(L)
        out = comp_downs_ref(x, 2)
        assert len(out) == int(np.ceil(L / 2))

    def test_skip_parameter(self):
        """skip=3: first output is x[3]."""
        x    = np.arange(10, dtype=float)
        out  = comp_downs_ref(x, 2, skip=3)
        assert out[0] == x[3]


# ---------------------------------------------------------------------------
# comp_ups – reference tests
# ---------------------------------------------------------------------------

class TestCompUpsReference:
    """
    MATLAB counterpart: TestSubsampling (comp_ups section).
    """

    def test_identity_a1(self):
        """comp_ups(x, 1) == x."""
        rng = np.random.default_rng(3)
        for _ in range(10):
            x = rng.standard_normal(128)
            np.testing.assert_array_equal(comp_ups_ref(x, 1), x)

    def test_output_length(self):
        """comp_ups(x, a) has length a * len(x)."""
        rng = np.random.default_rng(4)
        for _ in range(10):
            x = rng.standard_normal(64)
            a = int(rng.integers(2, 9))
            assert len(comp_ups_ref(x, a)) == a * len(x)

    def test_energy_preserved(self):
        """Upsampling preserves energy: sum(|up|^2) == sum(|x|^2)."""
        rng = np.random.default_rng(5)
        for _ in range(10):
            x   = rng.standard_normal(64)
            a   = int(rng.integers(2, 9))
            up  = comp_ups_ref(x, a)
            np.testing.assert_allclose(
                np.sum(np.abs(up) ** 2), np.sum(np.abs(x) ** 2), rtol=1e-12
            )

    def test_cyclic_impulse_structure(self):
        """After downsample + upsample the impulse is at position 0."""
        x = np.zeros(128); x[0] = 1.0
        down = comp_downs_ref(x, 2)
        up   = comp_ups_ref(down, 2)
        assert up[0] > 0

